File: python/mirage/commands/builtin/generic/fmt.py
import textwrap
from collections.abc import AsyncIterator, Awaitable, Callable

from mirage.commands.builtin.utils.stream import _read_stdin_async
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _fmt_text(text: str, width: int) -> str:
    paragraphs = text.split("\n\n")
    formatted = []
    for para in paragraphs:
        para = para.strip()
        if para:
            formatted.append(textwrap.fill(para, width=width))
        else:
            formatted.append("")
    return "\n\n".join(formatted) + "\n"


async def fmt(
    paths: list[PathSpec],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    accessor: object = None,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    width: int = 75,
) -> tuple[ByteSource | None, IOResult]:
    if paths:
        all_text: list[str] = []
        for p in paths:
            data = (await read_bytes(accessor, p)).decode(errors="replace")
            all_text.append(data)
        return _fmt_text("".join(all_text), width).encode(), IOResult()

    raw = await _read_stdin_async(stdin)
    if raw is None:
        raise ValueError("fmt: missing operand")
    text = raw.decode(errors="replace")
    return _fmt_text(text, width).encode(), IOResult()


__all__ = ["fmt"]

File: typescript/packages/browser/src/commands/builtin/opfs/fmt.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  IOResult,
  ResourceName,
  command,
  materialize,
  readStdinAsync,
  specOf,
  type ByteSource,
  type CommandFnResult,
  type CommandOpts,
  type PathSpec,
} from '@struktoai/mirage-core'
import { stream as opfsStream } from '../../../core/opfs/stream.ts'
import type { OPFSAccessor } from '../../../accessor/opfs.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

function wrapText(text: string, width: number): string {
  const words = text.split(/\s+/).filter((w) => w !== '')
  if (words.length === 0) return ''
  const lines: string[] = []
  let current = ''
  for (const w of words) {
    if (current === '') {
      current = w
    } else if (current.length + 1 + w.length <= width) {
      current = current + ' ' + w
    } else {
      lines.push(current)
      current = w
    }
  }
  if (current !== '') lines.push(current)
  return lines.join('\n')
}

function fmtText(text: string, width: number): string {
  const paragraphs = text.split('\n\n')
  const formatted: string[] = []
  for (const para of paragraphs) {
    const trimmed = para.trim()
    if (trimmed !== '') formatted.push(wrapText(trimmed, width))
    else formatted.push('')
  }
  return formatted.join('\n\n') + '\n'
}

async function fmtCommand(
  accessor: OPFSAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const width = typeof opts.flags.w === 'string' ? Number.parseInt(opts.flags.w, 10) : 75
  if (paths.length > 0) {
    const parts: string[] = []
    for (const p of paths) {
      parts.push(DEC.decode(await materialize(opfsStream(accessor, p))))
    }
    const result: ByteSource = ENC.encode(fmtText(parts.join(''), width))
    return [result, new IOResult()]
  }
  const stdinData = await readStdinAsync(opts.stdin)
  if (stdinData === null) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('fmt: missing operand\n') })]
  }
  const text = DEC.decode(stdinData)
  const result: ByteSource = ENC.encode(fmtText(text, width))
  return [result, new IOResult()]
}

export const OPFS_FMT = command({
  name: 'fmt',
  resource: ResourceName.OPFS,
  spec: specOf('fmt'),
  fn: fmtCommand,
})
